File: Factorising_algs/trails_division/trails.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>
#include <time.h>
#include <stdint.h>

int is_prime(unsigned long long n);
unsigned long long factorize(unsigned long long n);

void main(){
	printf("TRAILS IN C\n");
    unsigned long long p,q,n,x,y;
    clock_t begin = clock();


	p = 68121853;
	q = 97023187;
    n = p*q;
    x = factorize(n);
    y = n/x;
    printf("n = %lld\np = %lld, prime = %d\nq = %lld, prime = %d\n", n, x, is_prime(x), y, is_prime(y));
    if (x*y == n){
        printf("p*q does equal n\n");
    }
    clock_t end = clock();
    double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;
    printf("time taken = %.3f", time_spent);
}

int is_prime(unsigned long long n){
    if (n <= 1){
        return 0;
    }
    if (n == 2){
        return 1;
    }
    unsigned long long usable_root_n = sqrt(n)+1;
    if (n%2 == 0){
        return 0;
    }
    for (unsigned long long p=3; p < usable_root_n; p = p+2){
        if (n%p == 0){
            return 0;
        }
    }
    return 1;
}

unsigned long long factorize(unsigned long long n){ // this only factorizes if number found is prime!!
    if (n <= 2){
        return 0;
    }
    unsigned long long usable_root_n = sqrt(n)+1;
    if (n%2 == 0){
        return 2;
    }
    for (unsigned long long p=3; p < usable_root_n; p = p+2){
        if (n%p == 0 && is_prime(p)){
            return p;
        }
    }
    return 0;
}

File: Factorising_algs/trails_division/trails.py
from math import sqrt
#import sympy
#import random
import rsa 
import timeit

# O(n**(1/2))

def is_prime(n):
    if n <= 1:
        return 0
    if n == 2:
        return 1
    if n%2 == 0:
        return 0                                                 # every even number doesnt have to be tested if 2 isnt a factor
    for p in range(3,int(sqrt(n))+1,2):
        if n%p == 0:
            return 0
    return 1

# my implementation
def factorize(n):
    if n <= 1:
        return 0
    if n%2 == 0:
        return 2                                                  # every even number doesnt have to be tested if 2 isnt a factor
    for p in range(3,int(sqrt(n))+1,2):
        if n%p == 0 and is_prime(p):
            return p
    return 0

if __name__ == "__main__":
    print("TRAILS IN PYTHON")
    bits = 55
    num_of_runs = 1   
    print(f"num of runs per alg = {num_of_runs}, num of bits in n = {bits}")
    def test():
        p,q = rsa.prime.getprime(bits // 2), rsa.prime.getprime(bits // 2)
        n = p*q
        x = factorize(n)
        y = int(n/x)
        #print(f"n = {n}\np = {p}, prime = {is_prime(p)}\nq = {q}, prime = {is_prime(q)}")
        if x*y == n:
            print("p*q does equal n")

    time_time = timeit.Timer(test).timeit(number=num_of_runs)
    print(f"Average time of {round(time_time/num_of_runs,4)}s for {bits} bit N")
